src/bsd/mod.rs

```diff
@@ -0,0 +1,225 @@
+        IpAddr::V4(ip) => {
+            // Reinterpret the `sockaddr_storage` as `sockaddr_in`.
+            let sin = unsafe { &mut *ptr::from_mut(&mut dst).cast::<sockaddr_in>() };
+            sin.sin_len = size_of::<sockaddr_in>() as u8;
+            sin.sin_family = AF_INET_U8;
+            sin.sin_addr.s_addr = u32::from_ne_bytes(ip.octets());
+        }
+        #[allow(clippy::cast_possible_truncation)] // Guarded by `const_assert!` above.
+        IpAddr::V6(ip) => {
+            // Reinterpret the `sockaddr_storage` as `sockaddr_in6`.
+            let sin6 = unsafe { &mut *ptr::from_mut(&mut dst).cast::<sockaddr_in6>() };
+            sin6.sin6_len = size_of::<sockaddr_in6>() as u8;
+            sin6.sin6_family = AF_INET6_U8;
+            sin6.sin6_addr.s6_addr = ip.octets();
+        }
+    };
+    dst
+}
+
+pub fn if_index(remote: IpAddr) -> Result<u16, Error> {
+    // Open route socket.
+    let fd = unsafe { socket(PF_ROUTE, SOCK_RAW, AF_UNSPEC) };
+    if fd == -1 {
+        return Err(Error::last_os_error());
+    }
+    // Let OwnedFd take care of closing the file descriptor.
+    let fd = unsafe { OwnedFd::from_raw_fd(fd) };
+
+    // Prepare buffer with destination `sockaddr`.
+    let dst = as_sockaddr_storage(remote);
+
+    // Prepare route message structure.
+    let mut query: rt_msghdr = unsafe { zeroed() };
+    #[allow(clippy::cast_possible_truncation)]
+    // Structs len is <= u8::MAX per `const_assert!`s above; `aligned_by` returns max. 16 for IPv6.
+    let rtm_msglen = (size_of::<rt_msghdr>() + aligned_by(dst.ss_len.into(), ALIGN)) as u16; // Length includes sockaddr
+    query.rtm_msglen = rtm_msglen;
+    query.rtm_version = RTM_VERSION_U8;
+    query.rtm_type = RTM_GET_U8;
+    query.rtm_seq = fd.as_raw_fd(); // Abuse file descriptor as sequence number, since it's unique
+    query.rtm_addrs = RTM_ADDRS;
+    #[cfg(target_os = "openbsd")]
+    {
+        query.rtm_hdrlen = size_of::<rt_msghdr>() as libc::c_ushort;
+    }
+
+    // Copy route message and destination `sockaddr` into message buffer.
+    let mut msg: Vec<u8> = vec![0; query.rtm_msglen.into()];
+    unsafe {
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&query).cast(),
+            msg.as_mut_ptr(),
+            size_of::<rt_msghdr>(),
+        );
+        ptr::copy_nonoverlapping(
+            ptr::from_ref(&dst).cast(),
+            msg.as_mut_ptr()
+                .add(aligned_by(size_of::<rt_msghdr>(), ALIGN)),
+            dst.ss_len.into(),
+        );
+    }
+
+    // Send route message.
+    let res = unsafe { write(fd.as_raw_fd(), msg.as_ptr().cast(), msg.len()) };
+    if res == -1 {
+        return Err(Error::last_os_error());
+    }
+
+    // Read route messages.
+    let mut buf = vec![
+        0u8;
+        size_of::<rt_msghdr>() +
+        // There will never be `RTAX_MAX` sockaddrs attached, but it's a safe upper bound.
+         (RTAX_MAX as usize * size_of::<sockaddr_storage>())
+    ];
+    loop {
```

Comment:
If the message we `write` is well-formed, the kernel should always provide a response we can `read`. If the message isn't, the kernel should always respond with an error on the write. So the loop  _should_ always terminate.

I could of course switch `read` to `recv` and add a timeout?

Reply:
No - using `read` should be fine. It would probably be good to document the safety of this operation (i.e. that we're expecting the kernel to be well-behaved (which is a reasonable assumption, because if it isn't, then all bets are off)).